scripts/solve_network.py

```diff
@@ -1273,129 +1273,125 @@ def check_objective_value(n: pypsa.Network, solving: dict) -> None:
-        kwargs["remote"] = oetc_handler
-
-    kwargs["model_kwargs"] = cf_solving.get("model_kwargs", {})
-    kwargs["keep_files"] = cf_solving.get("keep_files", False)
+        solve_kwargs["remote"] = oetc_handler
 
-    if kwargs["solver_name"] == "gurobi":
+    if solver_name == "gurobi":
         logging.getLogger("gurobipy").setLevel(logging.CRITICAL)
 
-    rolling_horizon = cf_solving.pop("rolling_horizon", False)
-    skip_iterations = cf_solving.pop("skip_iterations", False)
-    if not n.lines.s_nom_extendable.any():
-        skip_iterations = True
-        logger.info("No expandable lines found. Skipping iterative solving.")
+    return model_kwargs, solve_kwargs
+
+
+def prepare_optimization_problem(
+    n: pypsa.Network,
+    config: dict,
+    params: dict,
+    solving: dict,
+    planning_horizons: str | None = None,
+    **model_kwargs_override,
+) -> tuple[dict, dict]:
+    """
+    Prepare optimization problem by creating model and adding extra functionality.
+
+    This function:
+    1. Attaches config and params to network for extra_functionality
+    2. Collects model and solver kwargs
+    3. Creates the optimization model
+    4. Adds extra functionality (custom constraints)
+
+    Parameters
+    ----------
+    n : pypsa.Network
+        The PyPSA network instance
+    config : dict
+        Configuration dictionary containing solver settings
+    params : dict
+        Dictionary of solving parameters
+    solving : dict
+        Dictionary of solving options and configuration
+    planning_horizons : str, optional
+        The current planning horizon year or None in perfect foresight
+    **model_kwargs_override
+        Override specific model_kwargs (e.g., log_fn)
 
-    # add to network for extra_functionality
+    Returns
+    -------
+    tuple[dict, dict]
+        (model_kwargs, solve_kwargs) used for problem creation
+    """
+    # Add config and params to network for extra_functionality
     n.config = config
     n.params = params
 
-    if rolling_horizon and rule_name == "solve_operations_network":
-        kwargs["horizon"] = cf_solving.get("horizon", 365)
-        kwargs["overlap"] = cf_solving.get("overlap", 0)
-        n.optimize.optimize_with_rolling_horizon(**kwargs)
-        status, condition = "", ""
-    elif skip_iterations:
-        status, condition = n.optimize(**kwargs)
-    else:
-        kwargs["track_iterations"] = cf_solving["track_iterations"]
-        kwargs["min_iterations"] = cf_solving["min_iterations"]
-        kwargs["max_iterations"] = cf_solving["max_iterations"]
-        if cf_solving["post_discretization"].pop("enable"):
-            logger.info("Add post-discretization parameters.")
-            kwargs.update(cf_solving["post_discretization"])
-        status, condition = n.optimize.optimize_transmission_expansion_iteratively(
-            **kwargs
-        )
+    # Prepare solver kwargs
+    model_kwargs, solve_kwargs = collect_kwargs(config, solving, planning_horizons)
```

Comment:
It would be even better for modularity and debugging if this function were not called inside `prepare_optimization_problem` but called outside with `model_kwargs` and `solve_kwargs` passed as arguments.

```py
 model_kwargs, solve_kwargs = collect_kwargs(config, solving, planning_horizons)
 prepare_optimization_problems(..., model_kwargs, solve_kwargs)
 ```